lighthouse-core/audits/treemap-data.js

```diff
@@ -0,0 +1,315 @@
+ * @typedef RootNode
+ * @property {string} name Arbitrary name identifier. Usually a script url.
+ * @property {Node} node
+ */
+
+/**
+ * @typedef Node
+ * @property {string} name Arbitrary name identifier. Usually a path component from a source map.
+ * @property {number} resourceBytes
+ * @property {number=} unusedBytes
+ * @property {string=} duplicate If present, this module is a duplicate. String is normalized source path. See ModuleDuplication.normalizeSource
+ * @property {Node[]=} children
+ */
+
+/**
+ * @typedef {Omit<Node, 'name'|'children'>} SourceData
+ */
+
+class TreemapDataAudit extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'treemap-data',
+      scoreDisplayMode: Audit.SCORING_MODES.INFORMATIVE,
+      title: 'Treemap Data',
+      description: 'Used for treemap app',
+      requiredArtifacts:
+        ['traces', 'devtoolsLogs', 'SourceMaps', 'ScriptElements', 'JsUsage', 'URL'],
+    };
+  }
+
+  /**
+   * Returns a tree data structure where leaf nodes are sources (ie. real files from source tree)
+   * from a source map, and non-leaf nodes are directories. Leaf nodes have data
+   * for bytes, coverage, etc., when available, and non-leaf nodes have the
+   * same data as the sum of all descendant leaf nodes.
+   * @param {string} sourceRoot
+   * @param {Record<string, SourceData>} sourcesData
+   * @return {Node}
+   */
+  static prepareTreemapNodes(sourceRoot, sourcesData) {
+    /**
+     * @param {string} name
+     * @return {Node}
+     */
+    function newNode(name) {
+      return {
+        name,
+        resourceBytes: 0,
+      };
+    }
+
+    /**
+     * Given a slash-delimited path, traverse the Node structure and increment
+     * the data provided for each node in the chain. Creates nodes as needed.
+     * Ex: path/to/file.js will find or create "path" on `node`, increment the data fields,
+     *     and continue with "to", and so on.
+     * @param {string} source
+     * @param {SourceData} data
+     * @param {Node} node
+     */
+    function addNode(source, data, node) {
+      // Strip off the shared root.
+      const sourcePathSegments = source.replace(sourceRoot, '').split(/\/+/);
+      sourcePathSegments.forEach((sourcePathSegment, i) => {
+        const isLastSegment = i === sourcePathSegments.length - 1;
+
+        let child = node.children && node.children.find(child => child.name === sourcePathSegment);
+        if (!child) {
+          child = newNode(sourcePathSegment);
+          node.children = node.children || [];
+          node.children.push(child);
+        }
+        node = child;
+
+        // Now that we've found or created the next node in the path, apply the data.
+        node.resourceBytes += data.resourceBytes;
+        if (data.unusedBytes) node.unusedBytes = (node.unusedBytes || 0) + data.unusedBytes;
+        if (data.duplicate !== undefined && isLastSegment) {
```

Comment:
why can only the leaf nodes be duplicates can't a module folder be duplicated? I'm looking for a comment in here if it's true :)

Reply:
from below it looks like our current code only is able to detect duplicate modules, which seems fine I just wouldn't expect to need this condition here then

Reply:
> Is it worth having for readability purposes?

> I'm looking for a comment in here if it's true :)

sure :)